betty/plugin/__init__.py

```diff
@@ -46,68 +49,108 @@ class PluginError(Exception):
     """
 
 
-class Plugin(ABC):
+class PluginDefinition:
     """
-    A plugin.
-
-    Plugins are identified by their :py:meth:`IDs <betty.plugin.Plugin.plugin_id>` as well as their types.
-    Each must be able to uniquely identify the plugin within a plugin repository.
-
-    To test your own subclasses, use :py:class:`betty.test_utils.plugin.PluginTestBase`.
+    A plugin definition.
     """
 
-    @classmethod
-    @abstractmethod
-    def plugin_type_cls(cls) -> type[Plugin]:
-        """
-        The base type (class) of all plugins of this type.
-        """
-
-    @classmethod
-    @abstractmethod
-    def plugin_type_id(cls) -> MachineName:
-        """
-        The plugin type ID.
-        """
-
-    @classmethod
-    @abstractmethod
-    def plugin_type_label(cls) -> Localizable:
-        """
-        Get the human-readable short plugin type label for the given count.
-        """
+    def __init__(
+        self,
+        *,
+        id: MachineName,  # noqa: A002
+    ):
+        self._id = id
 
-    @classmethod
-    @abstractmethod
-    def plugin_id(cls) -> MachineName:
+    @property
+    def id(self) -> MachineName:
         """
-        Get the plugin ID.
+        The plugin ID.
 
         IDs are unique per plugin type:
 
         - A plugin repository **MUST** at most have a single plugin for any ID.
         - Different plugin repositories **MAY** each have a plugin with the same ID.
         """
+        return self._id
 
-    @classmethod
-    @abstractmethod
-    def plugin_label(cls) -> Localizable:
+
+_PluginDefinitionT = TypeVar("_PluginDefinitionT", bound=PluginDefinition)
+
+
+class UserFacingPluginDefinition(UserFacing, PluginDefinition):
+    """
+    A definition of a plugin that is user-facing.
+    """
+
+    def __init__(
+        self,
+        *,
+        label: Localizable,
+        label_plural: Localizable,
```

Comment:
Extract countability into a separate type?

Reply:
Or pluralization, specifically. Also find a solution for defining countable plurals.